modules/git/commit_reader.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package git
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"strings"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
+
+// CommitFromReader will generate a Commit from a provided reader
+// We will need this to interpret commits from cat-file
+func CommitFromReader(gitRepo *Repository, sha plumbing.Hash, reader io.Reader) (*Commit, error) {
+	commit := &Commit{
+		ID: sha,
+	}
+
+	payloadSB := new(strings.Builder)
+	signatureSB := new(strings.Builder)
+	messageSB := new(strings.Builder)
+	message := false
+	pgpsig := false
+
+	scanner := bufio.NewScanner(reader)
```

Comment:
Git uses LF as the delimiter in the commit object format.

CR/LF could only appear in the message.

Reply:
Rather annoyingly my phone lost the second more complex reply to this...

In terms of "binary" the git commit object format is relatively undefined but the payload in the case of a commit is the commit, less the gpgsign element.

The code within git that does the parsing splitting the commit into its signature and payload is:

https://github.com/git/git/blob/228f53135a4a41a37b6be8e4d6e2b6153db4a8ed/commit.c#L1005-L1043

My code follows the pattern in go-git commit.Decode:

https://github.com/src-d/go-git/blob/1a7db85bca7027d90afdb5ce711622aaac9feaed/plumbing/object/commit.go#L160-L231

If you look closely there is a slight discrepancy in the way these two things work. Go-Git needs to move to be closer to the git code here - but I'm not sure if I should just keep things in the go-git way rather than move to the git way.

There's also the issue that everywhere in Gitea we make the assumption that everything git will give is UTF8 which is wrong. 

Now refactoring that is way out of scope for this PR. I'm happy to do that but we're talking about 1000s of lines of code.

Reply:
Oh damn I see.

Reply:
Done